Don't round microseconds when there aren't any.

petejohansonxo opened the pull request:
* When dealing w/ Time instances that have no microseconds,
  skip doing any precision logic on them.

Sample benchmark results:

===================== Time - No Precision, No Milliseconds =====================

Warming up --------------------------------------
                old?    24.642k i/100ms
                new?   117.499k i/100ms
Calculating -------------------------------------
                old?    276.648k (±21.8%) i/s -      1.306M in   5.023403s
                new?      1.644M (±19.2%) i/s -      7.872M in   5.013681s

Comparison:
                new?:  1643874.2 i/s
                old?:   276647.7 i/s - 5.94x  slower

================== Time - No Precision, Includes Milliseconds ==================

Warming up --------------------------------------
                old?    23.376k i/100ms
                new?    25.212k i/100ms
Calculating -------------------------------------
                old?    285.805k (±21.3%) i/s -      1.356M in   5.022266s
                new?    301.691k (±19.3%) i/s -      1.462M in   5.075615s

Comparison:
                new?:   301690.7 i/s
                old?:   285805.1 i/s - same-ish: difference falls within error

================== Time - Include Precision, No Milliseconds ===================

Warming up --------------------------------------
                old?    26.540k i/100ms
                new?   123.456k i/100ms
Calculating -------------------------------------
                old?    292.486k (±28.0%) i/s -      1.274M in   5.048784s
                new?      1.789M (±12.7%) i/s -      8.889M in   5.065824s

Comparison:
                new?:  1788969.3 i/s
                old?:   292486.1 i/s - 6.12x  slower

=============== Time - Include Precision, Includes Milliseconds ================

Warming up --------------------------------------
                old?    20.501k i/100ms
                new?    18.281k i/100ms
Calculating -------------------------------------
                old?    207.049k (±19.8%) i/s -    984.048k in   5.025614s
                new?    211.098k (±16.9%) i/s -      1.024M in   5.031450s

Comparison:
                new?:   211098.2 i/s
                old?:   207049.5 i/s - same-ish: difference falls within error

### Summary

When doing some work in our API, which uses Sinatra + ActiveRecord, In one of our "data-heavy" API endpoints, when using rack-mini-profiler and inspecting the flamegraph, I noticed a curious number of samples spent in `ActiveModel::Type::Helpers::TimeValue#apply_seconds_precision`, in many of them about 10% of the time for that call.

In our case, we are using MySQL 5.6.10 database, with default (0) precision datetime/timestamp columns, so for those the precision is `0`, and all of our values we are pulling out of the *database* have `0` usec values.

Given this, the code in `apply_seconds_precision` was still trying to do the calculations and update the `usec` value from.... `0` to `0`!

This one line change I'm proposing checks for this edge case in the existing guard block and adds one more condition for the early return.

### Other Information

Sample benchmark run was included in my commit message, here is the benchmark itself for posterity:

```ruby
# frozen_string_literal: true

require "bundler/inline"

gemfile(true) do
  source "https://rubygems.org"

  git_source(:github) { |repo| "https://github.com/#{repo}.git" }

  gem "rails", github: "rails/rails"
  gem "benchmark-ips"
end

require "active_model"

# Current code for ActiveModel::Type::Helpers::TimeValue#apply_seconds_precision
def apply_precision_old(precision, value)
  return value unless precision && value.respond_to?(:usec)
  number_of_insignificant_digits = 6 - precision
  round_power = 10**number_of_insignificant_digits
  value.change(usec: value.usec - value.usec % round_power)
end

# Propposed code for ActiveModel::Type::Helpers::TimeValue#apply_seconds_precision
# that checks for positive usec before doing further calculations
def apply_precision_new(precision, value)
  return value unless precision && value.respond_to?(:usec) && value.usec.positive?
  number_of_insignificant_digits = 6 - precision
  round_power = 10**number_of_insignificant_digits
  value.change(usec: value.usec - value.usec % round_power)
end

# Enumerate some representative scenarios here.
#
# It is very easy to make an optimization that improves performance for a
# specific scenario you care about but regresses on other common cases.
# Therefore, you should test your change against a list of representative
# scenarios. Ideally, they should be based on real-world scenarios extracted
# from production applications.
SCENARIOS = {
  "Time - No Precision, No Milliseconds"            => [0, Time.utc(2019,1,1,12,0,1)],
  "Time - No Precision, Includes Milliseconds"      => [0, Time.utc(2019,1,1,12,0,1,5123)],
  "Time - Include Precision, No Milliseconds"       => [6, Time.utc(2019,1,1,12,0,1)],
  "Time - Include Precision, Includes Milliseconds" => [6, Time.utc(2019,1,1,12,0,1,5123)],
}

SCENARIOS.each_pair do |name, args|
  puts
  puts " #{name} ".center(80, "=")
  puts

  precision, time = args

  Benchmark.ips do |x|
    x.report("old?") { apply_precision_old(precision, time.dup) }
    x.report("new?") { apply_precision_new(precision, time.dup) }
    x.compare!
  end
end
```

Are there other scenarios I should be benchmarking?

Thanks in advance.

petejohansonxo committed:
Don't round microseconds when there aren't any.

* When dealing w/ Time instances that have no microseconds,
  skip doing any precision logic on them.

Sample benchmark results:

===================== Time - No Precision, No Milliseconds =====================

Warming up --------------------------------------
                old?    24.642k i/100ms
                new?   117.499k i/100ms
Calculating -------------------------------------
                old?    276.648k (±21.8%) i/s -      1.306M in   5.023403s
                new?      1.644M (±19.2%) i/s -      7.872M in   5.013681s

Comparison:
                new?:  1643874.2 i/s
                old?:   276647.7 i/s - 5.94x  slower

================== Time - No Precision, Includes Milliseconds ==================

Warming up --------------------------------------
                old?    23.376k i/100ms
                new?    25.212k i/100ms
Calculating -------------------------------------
                old?    285.805k (±21.3%) i/s -      1.356M in   5.022266s
                new?    301.691k (±19.3%) i/s -      1.462M in   5.075615s

Comparison:
                new?:   301690.7 i/s
                old?:   285805.1 i/s - same-ish: difference falls within error

================== Time - Include Precision, No Milliseconds ===================

Warming up --------------------------------------
                old?    26.540k i/100ms
                new?   123.456k i/100ms
Calculating -------------------------------------
                old?    292.486k (±28.0%) i/s -      1.274M in   5.048784s
                new?      1.789M (±12.7%) i/s -      8.889M in   5.065824s

Comparison:
                new?:  1788969.3 i/s
                old?:   292486.1 i/s - 6.12x  slower

=============== Time - Include Precision, Includes Milliseconds ================

Warming up --------------------------------------
                old?    20.501k i/100ms
                new?    18.281k i/100ms
Calculating -------------------------------------
                old?    207.049k (±19.8%) i/s -    984.048k in   5.025614s
                new?    211.098k (±16.9%) i/s -      1.024M in   5.031450s

Comparison:
                new?:   211098.2 i/s
                old?:   207049.5 i/s - same-ish: difference falls within error

kamipo commented on activemodel/lib/active_model/type/helpers/time_value.rb:
```diff
@@ -22,7 +22,7 @@ def serialize(value)
 
         def apply_seconds_precision(value)
-          return value unless precision && value.respond_to?(:usec)
+          return value unless precision && value.respond_to?(:usec) && value.usec.positive?
```
The most slowest method call is the `value.change`.
We could also safe to bypass the line if `precision == 6`.

```ruby
if precision && precision < 6 && value.respond_to?(:usec) && value.usec.positive?
  ...
else
  value
end
```

petejohansonxo replied:
Indeed. Updated w/ the additional check for precision >= 6. Here is the updated benchmark used to run the results in my second commit:

```ruby
# frozen_string_literal: true

require "bundler/inline"

gemfile(true) do
  source "https://rubygems.org"

  git_source(:github) { |repo| "https://github.com/#{repo}.git" }

  gem "rails", github: "rails/rails"
  gem "benchmark-ips"
end

require "active_model"

# Current code for ActiveModel::Type::Helpers::TimeValue#apply_seconds_precision
def apply_precision_old(precision, value)
  return value unless precision && value.respond_to?(:usec)
  number_of_insignificant_digits = 6 - precision
  round_power = 10**number_of_insignificant_digits
  value.change(usec: value.usec - value.usec % round_power)
end

# Propposed code for ActiveModel::Type::Helpers::TimeValue#apply_seconds_precision
# that checks for positive usec before doing further calculations
def apply_precision_new(precision, value)
  return value unless precision && precision < 6 && value.respond_to?(:usec) && value.usec.positive?
  number_of_insignificant_digits = 6 - precision
  round_power = 10**number_of_insignificant_digits
  value.change(usec: value.usec - value.usec % round_power)
end

# Enumerate some representative scenarios here.
#
# It is very easy to make an optimization that improves performance for a
# specific scenario you care about but regresses on other common cases.
# Therefore, you should test your change against a list of representative
# scenarios. Ideally, they should be based on real-world scenarios extracted
# from production applications.
SCENARIOS = {
  "Time - No Precision, No Milliseconds"            => [0, Time.utc(2019,1,1,12,0,1)],
  "Time - No Precision, Includes Milliseconds"      => [0, Time.utc(2019,1,1,12,0,1,5123)],
  "Time - Include Mid Precision, No Milliseconds"       => [3, Time.utc(2019,1,1,12,0,1)],
  "Time - Include Mid Precision, Includes Milliseconds" => [3, Time.utc(2019,1,1,12,0,1,5123)],
  "Time - Include Max Precision, No Milliseconds"       => [6, Time.utc(2019,1,1,12,0,1)],
  "Time - Include Max Precision, Includes Milliseconds" => [6, Time.utc(2019,1,1,12,0,1,5123)],
}

SCENARIOS.each_pair do |name, args|
  puts
  puts " #{name} ".center(80, "=")
  puts

  precision, time = args

  Benchmark.ips do |x|
    x.report("old?") { apply_precision_old(precision, time.dup) }
    x.report("new?") { apply_precision_new(precision, time.dup) }
    x.compare!
  end
end
```

petejohansonxo committed:
Don't attempt to round usecs w/ max precision.

Updated benchmarks when skipping if precision >= 6:

===================== Time - No Precision, No Milliseconds =====================

Warming up --------------------------------------
                old?    28.251k i/100ms
                new?   126.202k i/100ms
Calculating -------------------------------------
                old?    317.724k (± 8.3%) i/s -      1.582M in   5.020115s
                new?      1.792M (±12.4%) i/s -      8.834M in   5.048513s

Comparison:
                new?:  1792412.2 i/s
                old?:   317724.0 i/s - 5.64x  slower

================== Time - No Precision, Includes Milliseconds ==================

Warming up --------------------------------------
                old?    28.931k i/100ms
                new?    27.980k i/100ms
Calculating -------------------------------------
                old?    318.343k (±16.5%) i/s -      1.562M in   5.078322s
                new?    313.981k (±16.1%) i/s -      1.539M in   5.067751s

Comparison:
                old?:   318342.6 i/s
                new?:   313981.3 i/s - same-ish: difference falls within error

================ Time - Include Mid Precision, No Milliseconds =================

Warming up --------------------------------------
                old?    28.175k i/100ms
                new?   124.067k i/100ms
Calculating -------------------------------------
                old?    327.630k (±16.3%) i/s -      1.606M in   5.070016s
                new?      1.829M (±12.1%) i/s -      8.933M in   4.999892s

Comparison:
                new?:  1829256.4 i/s
                old?:   327630.1 i/s - 5.58x  slower

============= Time - Include Mid Precision, Includes Milliseconds ==============

Warming up --------------------------------------
                old?    21.030k i/100ms
                new?    20.551k i/100ms
Calculating -------------------------------------
                old?    218.390k (± 7.8%) i/s -      1.094M in   5.037354s
                new?    223.061k (±14.8%) i/s -      1.110M in   5.104758s

Comparison:
                new?:   223061.0 i/s
                old?:   218390.4 i/s - same-ish: difference falls within error

================ Time - Include Max Precision, No Milliseconds =================

Warming up --------------------------------------
                old?    28.460k i/100ms
                new?   145.661k i/100ms
Calculating -------------------------------------
                old?    323.322k (±16.6%) i/s -      1.565M in   5.010460s
                new?      2.297M (±13.6%) i/s -     11.216M in   5.020398s

Comparison:
                new?:  2296764.8 i/s
                old?:   323321.9 i/s - 7.10x  slower

============= Time - Include Max Precision, Includes Milliseconds ==============

Warming up --------------------------------------
                old?    20.544k i/100ms
                new?   146.182k i/100ms
Calculating -------------------------------------
                old?    219.394k (± 6.6%) i/s -      1.109M in   5.080068s
                new?      2.261M (±13.5%) i/s -     11.110M in   5.052416s

Comparison:
                new?:  2260632.4 i/s
                old?:   219394.1 i/s - 10.30x  slower

kamipo commented on activemodel/lib/active_model/type/helpers/time_value.rb:
```diff
@@ -22,7 +22,7 @@ def serialize(value)
 
         def apply_seconds_precision(value)
-          return value unless precision && value.respond_to?(:usec)
+          return value unless precision && precision < 6 && value.respond_to?(:usec) && value.usec.positive?
```
This condition is now longer (complex) enough.
We generally use early returning only if simple condition.

```ruby
if precision && precision < 6 && value.respond_to?(:usec) && value.usec.positive?
  ...
else
  value
end
```

petejohansonxo replied:
Fixed. Thanks!

petejohansonxo committed:
Switch away from guard clause style.

kamipo commented:
umm... two tests added by 4d9126cfccefdb69149caf7681d674b50335e9b4 are failed, but that tests using `nsec` doesn't make sense to me.

Active Record only save/load `usec`, and also, no precision doesn't care about truncation but use natual `usec`. So max precision (= 6) regards as unlimited (no) precision is reasonable to me.

So can you change the tests using `nsec` to using `usec`?

cc @pixeltrix

petejohansonxo commented:
@kamipo Thinking about this more, I wonder if we still may need to apply the rounding w/ precision = 6, if this is also being used to round *input* values before they get persisted and reloaded to the database.

I know in the original PR that added the method in question, that was one of the target use cases.

kamipo commented:
IMO, I didn't intend to truncate `nsec` which isn't used in AR, that was to avoid the "rounding to 1 sec future" issue.
I'd not like to pay a cost to truncate the unused `nsec` (also, almost all commodity machines not have nsec precision: `Time.now.utc.nsec # => 468175000`).

petejohansonxo commented:
@kamipo Ok. I'm fine w/ adjusting the tests in question to usec. Will work on that. Thanks!

pixeltrix commented:
@kamipo the nsec tests were for the situation where the timestamps were changing after reloading and not rounding the nsec will change that behaviour. I'm unsure about your assertion that commodity hardware doesn't use nsec (I know macOS doesn't) - for example this is from a ordinary Digital Ocean droplet:

``` sh
irb(main):002:0> Time.now.nsec
=> 588671744
```

and this is from a EC2 instance:

``` sh
irb(main):001:0> Time.now.nsec
=> 571673596
```

Just worried that changing this behaviour could lead to subtle, hard to trace bugs in production.

petejohansonxo commented:
@pixeltrix Yeah, I was starting to consider the possible implications as well. I'm working on another commit that calculates the rounded usec, and only calls `value.update` if changed, to see if we can get some of the performance boosts, without the `precision < 6` check.

kamipo commented:
hmm... if so, only safer way is that bypassing the truncation if value.nsec == 0.

petejohansonxo committed:
Proper handling of nanosecond rounding.

New benchmark, including impact on times w/ nanoseconds:

===================== Time - No Precision, No Milliseconds =====================

Warming up --------------------------------------
                old?    27.767k i/100ms
                new?   128.564k i/100ms
Calculating -------------------------------------
                old?    327.115k (±16.6%) i/s -      1.583M in   5.013219s
                new?      1.839M (±12.6%) i/s -      8.999M in   5.021190s

Comparison:
                new?:  1839389.2 i/s
                old?:   327115.5 i/s - 5.62x  slower

================== Time - No Precision, Includes Milliseconds ==================

Warming up --------------------------------------
                old?    28.776k i/100ms
                new?    26.225k i/100ms
Calculating -------------------------------------
                old?    323.133k (±16.7%) i/s -      1.583M in   5.075262s
                new?    298.401k (±12.0%) i/s -      1.469M in   5.024835s

Comparison:
                old?:   323133.1 i/s
                new?:   298401.3 i/s - same-ish: difference falls within error

================== Time - No Precision, Includes Nanoseconds ===================

Warming up --------------------------------------
                old?    28.359k i/100ms
                new?    27.460k i/100ms
Calculating -------------------------------------
                old?    320.350k (±16.9%) i/s -      1.560M in   5.050734s
                new?    300.258k (± 8.2%) i/s -      1.510M in   5.068984s

Comparison:
                old?:   320349.8 i/s
                new?:   300258.2 i/s - same-ish: difference falls within error

================ Time - Include Mid Precision, No Milliseconds =================

Warming up --------------------------------------
                old?    29.082k i/100ms
                new?   124.161k i/100ms
Calculating -------------------------------------
                old?    327.657k (±16.6%) i/s -      1.600M in   5.054723s
                new?      1.827M (±13.1%) i/s -      8.940M in   5.034525s

Comparison:
                new?:  1827118.5 i/s
                old?:   327656.8 i/s - 5.58x  slower

============= Time - Include Mid Precision, Includes Milliseconds ==============

Warming up --------------------------------------
                old?    20.458k i/100ms
                new?    19.532k i/100ms
Calculating -------------------------------------
                old?    225.679k (±15.2%) i/s -      1.105M in   5.030362s
                new?    218.728k (±14.8%) i/s -      1.074M in   5.042616s

Comparison:
                old?:   225678.9 i/s
                new?:   218727.7 i/s - same-ish: difference falls within error

============== Time - Include Mid Precision, Includes Nanoseconds ==============

Warming up --------------------------------------
                old?    20.315k i/100ms
                new?    19.639k i/100ms
Calculating -------------------------------------
                old?    222.229k (±15.1%) i/s -      1.097M in   5.070824s
                new?    215.144k (±14.7%) i/s -      1.061M in   5.058222s

Comparison:
                old?:   222229.1 i/s
                new?:   215144.3 i/s - same-ish: difference falls within error

================ Time - Include Max Precision, No Milliseconds =================

Warming up --------------------------------------
                old?    29.194k i/100ms
                new?   122.246k i/100ms
Calculating -------------------------------------
                old?    315.993k (±19.0%) i/s -      1.547M in   5.123735s
                new?      1.703M (±12.8%) i/s -      8.435M in   5.067265s

Comparison:
                new?:  1703255.1 i/s
                old?:   315993.4 i/s - 5.39x  slower

============= Time - Include Max Precision, Includes Milliseconds ==============

Warming up --------------------------------------
                old?    20.182k i/100ms
                new?   106.505k i/100ms
Calculating -------------------------------------
                old?    221.495k (±14.2%) i/s -      1.090M in   5.039889s
                new?      1.378M (±10.4%) i/s -      6.816M in   5.016567s

Comparison:
                new?:  1377837.0 i/s
                old?:   221494.8 i/s - 6.22x  slower

============== Time - Include Max Precision, Includes Nanoseconds ==============

Warming up --------------------------------------
                old?    20.237k i/100ms
                new?   105.485k i/100ms
Calculating -------------------------------------
                old?    222.884k (±14.3%) i/s -      1.093M in   5.022628s
                new?      1.397M (±10.7%) i/s -      6.962M in   5.054807s

Comparison:
                new?:  1397256.9 i/s
                old?:   222884.1 i/s - 6.27x  slower

petejohansonxo commented:
@pixeltrix @kamipo Pushed a version w/ the conditional call to `value.change` and removing the `precision < 6` check, benchmarks seem to still indicate a decent speed up, with only minor slowness for the case where not having that extra check before "helped" that path.

Still messing with trying to make my system happy to install the libxml-ruby gem locally so I can run the tests locally here, unfortunately, but hopefull that the CI checks will pass w/ this version.

Thanks.

pixeltrix commented:
There is also the question over whether any typical real-world scenario would see a speed-up - looking at the benchmarks it takes around 5µs to do the truncation so even with 200 timestamps on a record you would only see a benefit of 1ms per query. It may add up in edge cases but it's inconsequential for most Rails apps. Not saying that we shouldn't merge it though - just that it's not worth making the code harder to understand so that we have a theoretical gain but no real-world benefit.

petejohansonxo commented:
@pixeltrix In our case, this is for a data heavy endpoint from our scheduling system, that's a join on several models to return a "fat API response" with a lot of date/time related fields, which is probably the worst case scenario and the only reason I spotted it. Certainly an edge case micro-optimization I happen to hit.

petejohansonxo committed:
Some test fixes.

kamipo commented:
I've exercised the difference between no precision and max precision b78e3b5e21b9924324c4391cc58afac8c7b4cccb.
For now this PR will break the existing behavior.

petejohansonxo committed:
One more attempt at test fixes.

petejohansonxo committed:
Tweaks for proper nsec handling.

Latest benchmarks:

===================== Time - No Precision, No Milliseconds =====================

Warming up --------------------------------------
                old?    29.237k i/100ms
                new?   112.490k i/100ms
Calculating -------------------------------------
                old?    333.789k (±16.0%) i/s -      1.637M in   5.067766s
                new?      1.528M (±10.2%) i/s -      7.649M in   5.072327s

Comparison:
                new?:  1527763.1 i/s
                old?:   333789.1 i/s - 4.58x  slower

================== Time - No Precision, Includes Milliseconds ==================

Warming up --------------------------------------
                old?    29.263k i/100ms
                new?    28.736k i/100ms
Calculating -------------------------------------
                old?    323.948k (±14.7%) i/s -      1.580M in   5.014251s
                new?    317.262k (±15.5%) i/s -      1.552M in   5.044623s

Comparison:
                old?:   323948.5 i/s
                new?:   317262.0 i/s - same-ish: difference falls within error

================== Time - No Precision, Includes Nanoseconds ===================

Warming up --------------------------------------
                old?    28.837k i/100ms
                new?    28.526k i/100ms
Calculating -------------------------------------
                old?    320.457k (±15.3%) i/s -      1.557M in   5.007947s
                new?    309.512k (±15.2%) i/s -      1.512M in   5.026113s

Comparison:
                old?:   320457.2 i/s
                new?:   309511.7 i/s - same-ish: difference falls within error

================ Time - Include Mid Precision, No Milliseconds =================

Warming up --------------------------------------
                old?    28.318k i/100ms
                new?   109.599k i/100ms
Calculating -------------------------------------
                old?    294.268k (±18.4%) i/s -      1.444M in   5.112594s
                new?      1.402M (±12.8%) i/s -      6.905M in   5.016563s

Comparison:
                new?:  1402352.7 i/s
                old?:   294268.4 i/s - 4.77x  slower

============= Time - Include Mid Precision, Includes Milliseconds ==============

Warming up --------------------------------------
                old?    17.349k i/100ms
                new?    17.261k i/100ms
Calculating -------------------------------------
                old?    222.787k (± 7.7%) i/s -      1.110M in   5.019197s
                new?    222.177k (±13.7%) i/s -      1.087M in   5.014105s

Comparison:
                old?:   222787.0 i/s
                new?:   222177.3 i/s - same-ish: difference falls within error

============== Time - Include Mid Precision, Includes Nanoseconds ==============

Warming up --------------------------------------
                old?    19.555k i/100ms
                new?    18.861k i/100ms
Calculating -------------------------------------
                old?    216.203k (± 9.3%) i/s -      1.076M in   5.021809s
                new?    208.062k (±14.6%) i/s -      1.018M in   5.018286s

Comparison:
                old?:   216202.7 i/s
                new?:   208061.8 i/s - same-ish: difference falls within error

================ Time - Include Max Precision, No Milliseconds =================

Warming up --------------------------------------
                old?    29.488k i/100ms
                new?   111.379k i/100ms
Calculating -------------------------------------
                old?    320.059k (±16.4%) i/s -      1.563M in   5.053817s
                new?      1.415M (±14.5%) i/s -      6.905M in   5.005373s

Comparison:
                new?:  1414919.4 i/s
                old?:   320058.6 i/s - 4.42x  slower

============= Time - Include Max Precision, Includes Milliseconds ==============

Warming up --------------------------------------
                old?    18.227k i/100ms
                new?    18.580k i/100ms
Calculating -------------------------------------
                old?    220.021k (± 8.3%) i/s -      1.094M in   5.008914s
                new?    219.922k (±13.9%) i/s -      1.078M in   5.012089s

Comparison:
                old?:   220020.7 i/s
                new?:   219921.7 i/s - same-ish: difference falls within error

============== Time - Include Max Precision, Includes Nanoseconds ==============

Warming up --------------------------------------
                old?    20.698k i/100ms
                new?    19.372k i/100ms
Calculating -------------------------------------
                old?    211.239k (± 8.8%) i/s -      1.056M in   5.039393s
                new?    209.694k (±14.3%) i/s -      1.027M in   5.016998s

Comparison:
                old?:   211239.1 i/s
                new?:   209694.3 i/s - same-ish: difference falls within error

petejohansonxo commented:
@kamipo @pixeltrix I was able to implement this in a a way that still preserved the previous behaviour for truncating nanoseconds, if needed. Still a speed up, and the code still seems fairly readable.

Thoughts?

kamipo commented on activemodel/lib/active_model/type/helpers/time_value.rb:
```diff
@@ -38,6 +42,11 @@ def user_input_in_time_zone(value)
 
+          def value_is_rounding_candidate?(value)
+            value.respond_to?(:usec) &&
+              (value.usec.positive? || value.respond_to?(:nsec) && (value.nsec % 1000).positive?)
```
Why just not use `value.nsec.positive?`?

petejohansonxo replied:
I suppose I could... For some reason I was thinking, since `nsec` contains the seconds/milliseconds total as well, not just the nanosecond remainder, that I should check that. But, that's an unnecessary calculation in the check, let me revise.

petejohansonxo committed:
Simplify check for whether to round a value.

kamipo commented on activemodel/lib/active_model/type/helpers/time_value.rb:
```diff
@@ -38,6 +42,11 @@ def user_input_in_time_zone(value)
 
+          def value_is_rounding_candidate?(value)
+            value.respond_to?(:usec) &&
+              (value.usec.positive? || value.respond_to?(:nsec) && value.nsec.positive?)
```
Checking `value.usec.positive?` is redundant, since if `value.usec.positive?` is true, `value.nsec.positive?` is also true.

petejohansonxo replied:
@kamipo I was trying to guard against the case we have a value that responds to `usec`, but *not* to `nsec`. If we don't need to handle that scenario, we certainly can just check `nsec.positive?` here.

pixeltrix replied:
I think everything responds to `nsec` now (in Rails at least - DateTime doesn't without Active Support loaded, but it also doesn't respond to `usec` either 😄 )

petejohansonxo commented:
Thanks!
